activerecord/lib/active_record/connection_adapters.rb

```diff
@@ -31,6 +31,68 @@ def resolve(adapter_name) # :nodoc:
         class_name, path_to_adapter = @adapters[adapter_name.to_s]
 
         unless class_name
+          # To provide better error messages for adapters expecting the pre-7.2 adapter registration API, we attempt
+          # to load the adapter file from the old location which was required by convention, and then raise an error
+          # describing how to upgrade the adapter to the new API.
+          legacy_adapter_path = "active_record/connection_adapters/#{adapter_name}_adapter"
+          legacy_adapter_connection_method_name = "#{adapter_name}_connection".to_sym
+          legacy_adapter_class_method_name = "#{adapter_name}_adapter_class".to_sym # sqlite3_adapter_class
+
+          begin
+            # Attempt to load the legacy adapter by the filename convention.
+            require legacy_adapter_path
+
+            # If it is loaded without raising we found the found a file that may be the legacy adapter.
+            if ActiveRecord::ConnectionHandling.instance_methods.include?(legacy_adapter_connection_method_name) &&
+              ActiveRecord::ConnectionHandling.instance_methods.include?(legacy_adapter_class_method_name) &&
```

Comment:
```suggestion
            if ActiveRecord::ConnectionHandling.method_defined?(legacy_adapter_connection_method_name) &&
              ActiveRecord::ConnectionHandling.method_defined?(legacy_adapter_class_method_name) &&
```

`instance_methods.include?` allocate a large array with all that class methods (including the inherited one) and then do an `O(n)` search into it.

`method_defined?` directly do a `O(1)` search in the method table.

Reply:
Fixed